server/src/storage/data.rs

```diff
@@ -41,27 +40,83 @@ impl<B: Write + Read + Seek> Rows <B> {
         size
     }
 
+    /// Returns the last not deleted row.
+    /// The search begins with the first row located before position.
+    /// An error is returned if no row could be found.
+    /// Moves cursor to the end of the read row.
+    fn get_last_row(&mut self, position: SeekFrom, ignore_deleted_rows: bool)
+        -> Result <Vec<u8>, Error>
+    {
+        try!(self.set_pos(position));
+        try!(self.prev_row());
+
+        // Moves cursor to the beginning of the first not deleted row
+        // from the end.
+        while ignore_deleted_rows && try!(self.is_next_row_deleted(true)) {
+            try!(self.prev_row());
+        }
+
+        let mut row_data = Vec::<u8>::new();
+        try!(self.next_row_ex(&mut row_data, false));
+        Ok(row_data)
+    }
+
+    /// Tests if the row after the cursor is marked as deleted
+    /// restore_cursor: if true, calling the function will move the cursor,
+    /// if false, the cursor will be moved by row_header.size() bytes.
+    fn is_next_row_deleted(&mut self, restore_cursor: bool)
+        -> Result <bool, Error>
+    {
+        info!("Reading Header");
+        let row_header: RowHeader = try!(self.read_header());
+        info!("Header Read");
+        if restore_cursor {
+            try!(self.set_pos(SeekFrom::Current(-(RowHeader::size() as i64))));
+        };
+        Ok(row_header.is_deleted())
+    }
+
+    /// Moves the cursor to the beginning of the next deleted row
+    fn move_to_next_deleted_row(&mut self) -> Result<(), Error> {
+        let mut row_data = Vec::<u8>::new();
+        while try!(self.is_next_row_deleted(true)) {
+            try!(self.next_row_ex(&mut row_data, false));
+        }
+        Ok(())
+    }
+
+    /// reads the next row, which is not marked as deleted
+    /// and writes the data into target_buf
+    /// returns the bytes read or an Error otherwise.
+    /// Returns Error:EndOfFile if no next row could be read.
+    pub fn next_row<W: Write>(&mut self,
+                              mut target_buf: &mut W)
+        -> Result<u64, Error>
+    {
+        self.next_row_ex(target_buf, true)
+    }
+
     /// reads the next row, which is not marked as deleted
     /// and writes the data into target_buf
     /// returns the bytes read or an Error otherwise.
     /// Returns Error:EndOfFile if no next row could be read.
-    pub fn next_row<W: Write>(&mut self, mut target_buf: &mut W)
+    /// if ignore_deleted_rows == true, deleteed rows will be skipped.
+    fn next_row_ex<W: Write>(&mut self,
+                              mut target_buf: &mut W,
```

Comment:
same as above.. but ok, if you like it more if every parameter is on its own line..